tests/Highlighter.py

```diff
@@ -0,0 +1,358 @@
+    "uncolor"             : "T",
+    "txt"                 : "T",
+    "white"               : "T",
+    "T"                   : "T",
+
+    "keyword"             : "K",
+    "kw"                  : "K",
+    "red"                 : "K",
+    "K"                   : "K",
+
+    "comment"             : "C",
+    "cmt"                 : "C",
+    "grey"                : "C",
+    "C"                   : "C",
+
+    "variable"            : "N",
+    "number"              : "N",
+    "green"               : "N",
+    "N"                   : "N",
+
+    "constant.character"  : "S",
+    "string"              : "S",
+    "str"                 : "S",
+    "blue"                : "S",
+    "S"                   : "S",
+
+    "invalid"             : "I",
+    "inv"                 : "I",
+    "pink"                : "I",
+    "I"                   : "I",
+}
+
+
+class HighlightTester(unittest.TestCase):
+
+    def assert_highlighted_chr(self, code, expected, level, lang="en"):
+        rules  = self.get_rules(level, lang)
+        self.check(code, expected, rules)
+
+
+    def assert_highlighted_chr_multi_line(self, *args, level, lang="en"):
+        if len(args)%2 != 0:
+            raise RuntimeError(f'Pass an even number of strings to assert_highlighted_chr_multi_line (alternatingly code and highlighting). Got: {args}')
+
+        code = []
+        expected = []
+        for i, line in enumerate(args):
+            if i%2 == 0: code.append(line)
+            else: expected.append(line)
+
+        code = "\n".join(code)
+        expected = "\n".join(expected)
+
+        self.assert_highlighted_chr(code, expected, level, lang)
+
+
+    def assert_highlighted(self, code_col, level, lang="en"):
+        code, expected = self.converts(code_col)
+        self.assert_highlighted_chr(code, expected, level, lang)
+
+
+    def assert_highlighted_multi_line(self, *args, level, lang="en"):
+        code_col = "\n".join(args)
+        code, expected = self.converts(code_col)
+        self.assert_highlighted_chr(code, expected, level, lang)
+
+
+
+    # apply rules on code and check if the result is valid
+    def check(self, code, expected, rules):
+        
+        simulator = SimulatorAce(rules)
+        result = simulator.highlight(code)
+
+        valid, ind_error = self.compare(result, expected)
+
+        if not valid:
+
+            code_list     = code.split('\n')
+            expected_list = expected.split('\n')
+            result_list   = result.split('\n')
+
+            line_cpt = 0
+            while ind_error >= len(code_list[line_cpt]):
+                ind_error -= len(code_list[line_cpt]) +1
+                line_cpt += 1
+
+            print("ERROR in this code :")
+            print("In this code :", code_list[line_cpt] )
+            print("We want      :", expected_list[line_cpt] )
+            print("We have      :", result_list[line_cpt] )
+            print("At           :", " " * ind_error + "^")
+
+        self.assertTrue(valid)
```

Comment:
In general: less code is better, because *code represents a cost*. The cost needs to be justified by the functionality it offers over readily available things.

By using `self.assertEquals` you could get rid of `self.compare` below.

Reply:
I close this channel because it is redundant with https://github.com/Felienne/hedy/pull/2474#discussion_r874728164